### ISSUE TITLE
Optimize Docker images using Spring Boot layered JAR

### ISSUE THREAD

**Issue by tonghoangvu** [nashtech-garage/yas#642]
The current approach for building Docker images is not really optimized.

Docker can't cache the JAR layer (2nd line), so every image contains a whole different JAR, even the smallest code change.

The Spring Boot JAR size is typically over 30MB, so each Docker image adds an unnecessary 30MB.

```Dockerfile
FROM eclipse-temurin:17-jre-alpine
ADD target/product-0.0.1-SNAPSHOT.jar product-0.0.1-SNAPSHOT.jar
ENTRYPOINT ["java", "-jar", "/product-0.0.1-SNAPSHOT.jar"]
```

Check out the link below for a better explanation.

https://www.baeldung.com/docker-layers-spring-boot

And here is the example optimized Dockerfile from one of my projects.

https://github.com/pocketo-app/pocketo-backend/blob/dev/Dockerfile

**Comment by duccaominh** [comment]
Please refer to this issue to find out more about this approach. We copy the jar from maven build stage, github CI/CD has the maven cache, which's more optimize when using with CI/CD flow. Multi stage docker build beside will be much more effective if we use standalone with `docker build` on local machine and run from that. Another approach that worth mentions here is Paketo cloud native buildpacks would achive the same thing but way more good in cloud native app
`https://github.com/nashtech-garage/yas/pull/16`

**Comment by tonghoangvu** [comment]
I see, but this is not about building source code inside the Docker build process. We still build the JAR outside (in the local machine or CI pipeline) and copy it to the Docker image.

![](https://www.baeldung.com/wp-content/uploads/2020/11/spring-boot-layers.jpg)

But with the new approach, we will extract the JAR file into multiple layers (folders) by how frequently it changes. Docker can now cache the unchanged layers, thus reducing the additional size with each version of the image.

For example, a small code change will only add a few KBs to the new Docker image (only the `application` layer changed). Compare this to the current approach, which has a single JAR layer only, which always adds an additional 50 MB at least.

![image](https://user-images.githubusercontent.com/64119341/233421086-0d4eb7f5-f08e-41b3-a338-97c11c18e946.png)

I found that Nashtech YAS is awesome, and I'm willing to submit a new pull request.

**Comment by duccaominh** [comment]
@tonghoangvu yeah i understand the point about layer jar, in the past i submitted a PR with share the same idea with you. But to keep this simple and focus more on business code, we're still using this approach for now 👍  nice work btw 💯

**Comment by tonghoangvu** [comment]
Nice, really appreciate it. Thank you.
